prunenn/adder.py

```diff
@@ -0,0 +1,105 @@
+import torch
+import torch.nn as nn
+from torch.nn.parameter import Parameter
+import torch.nn.functional as F
+import torch.optim as optim
+from torchvision import datasets, transforms
+import numpy as np
+from collections import OrderedDict
+import itertools
+
+'''
+Adds nodes to neural network
+'''
+class Adder(nn.Module):
+  def __init__(self, model, thres = 0.4, function = "var"):
+        super(Adder, self).__init__()
+        self.model = model
+        self.to_add = {}
+        self.next_layer = {}
+        self.thres = thres
+        self.to_train = True #true is train, false is test
+
+        layer_names = list(model.pruning_layers._modules.keys())
+        for i, layer_name in enumerate(layer_names):
+            layer = model.pruning_layers._modules.get(layer_name)
+            if(i < len(model.pruning_layers._modules)-1):
+                layer.layer_name = layer_name
+                self.next_layer[layer_name] = \
+                    model.pruning_layers._modules.get(layer_names[i+1])
+                if(function == "var" ): #and not to_train
+                    layer.register_forward_hook(self.low_variance_nodes)
+            i += 1
+
+  def add(self):
+      i = 0
+      for layer_name, seq_layer in model.pruning_layers._modules.items():
+        if(i < len(model.pruning_layers._modules)-1):
+          layer = seq_layer._modules['0']
+          #print("adding nodes ", layer_name, " ", self.to_add)
+          nodes_to_add = self.to_add[layer_name]
+          np_weights = layer.weight.data.cpu().numpy()
+
+          np_weights[nodes_to_add,:, :, :] = np.random.normal(np_weights[nodes_to_add,:, :, :], 0.5*np.std(np_weights[nodes_to_add,:, :, :]))
```

Comment:
the lack of spaces between the : after nodes_to_add is driving me crazy

Reply:
will fix